Harmony 943

vinnyinverso opened the pull request:
Cancel work items when a job fails.

bilts commented on app/backends/workflow-orchestration.ts:
```diff
@@ -159,16 +159,18 @@ export async function updateWorkItem(req: HarmonyRequest, res: Response): Promis
+        job.fail(message);
         try {
           await job.save(tx);
+          await updateWorkItemsStatusByJobId(
```
I'm wondering if this should all be taken a step further and put in a method on `job`.  Something like `job.complete(tx, finalStatus, message)`.  For that matter, I was a little surprised that this all didn't end up inside the `fail` method and that we have to save and update stuff after calling `fail`.  Maybe that's similar to your question from the stand up about the behavior of `save`.

Either way, I think a method on `Job` that knows all the non-terminal statuses could be helpful if we ever end up adding one

bilts commented on app/models/work-item.ts:
```diff
@@ -167,6 +167,31 @@ export async function updateWorkItemStatus(
+ * @param oldStatuses - restricts the updates to work items where the status is in oldStatuses
+ * @param newStatus - the value of the updated status
+ */
+export async function updateWorkItemsStatusByJobId(
```
Related to prior message, I could see this being a method on the job, i.e. `job.updateWorkItemStatuses`

bilts commented on test/workflow-orchestration.ts:
```diff
@@ -138,6 +138,70 @@ describe('Workflow chaining for a collection configured for swot reprojection an
+      });
+
+      describe('when the first swot-reprojection service work item fails', function () {
+        it('the job fails, and all further work items are canceled', async function () {
```
`it('fails the job and cancels all further work items...`

bilts commented on test/workflow-orchestration.ts:
```diff
@@ -138,6 +138,70 @@ describe('Workflow chaining for a collection configured for swot reprojection an
+          workItem.status = WorkItemStatus.FAILED;
+          workItem.results = [];
+          await updateWorkItem(this.backend, workItem);
+          expect(workItem.serviceID).to.equal('sds/swot-reproject:latest');
```
Here and above probably belong in a `before` block

bilts commented on app/util/job.ts:
```diff
@@ -38,11 +38,9 @@ export default async function cancelAndSaveJob(
-          .whereIn('status', [WorkItemStatus.READY, WorkItemStatus.RUNNING])
-          .update({ status: WorkItemStatus.CANCELED, updatedAt });
+        await updateWorkItemsStatusByJobId(
+          tx, job.jobID, [WorkItemStatus.READY, WorkItemStatus.RUNNING], WorkItemStatus.CANCELED,
```
I know this isn't quite your change, but if something is `RUNNING` and then gets `CANCELED` like this, presumably whatever call was working on it will call back and try to update its status.  Is that behavior handled properly?

vinnyinverso replied:
This is what happens in that case:

```
// If the job was already canceled then send 400 response
    if (job.status === JobStatus.CANCELED) {
      res.status(409).send('Job was already canceled.');
```

vinnyinverso replied:
Actually, the status of the work item would get updated again (undesirable) on the call back. I think I can just move the update invocation around to account for this.

bilts commented on test/workflow-orchestration.ts:
```diff
@@ -138,6 +138,70 @@ describe('Workflow chaining for a collection configured for swot reprojection an
+          const currentWorkItems = (await getWorkItemsByJobId(db, workItem.jobID)).workItems;
+          expect(currentWorkItems.length).to.equal(4);
+          expect(currentWorkItems.filter((i) => i.status === WorkItemStatus.SUCCESSFUL && i.serviceID === 'harmonyservices/query-cmr:latest').length).to.equal(1);
+          expect(currentWorkItems.filter((i) => i.status === WorkItemStatus.CANCELED && i.serviceID === 'sds/swot-reproject:latest').length).to.equal(2);
```
One way to guard against the issue I mentioned earlier, where we could add a status and it ends up not getting handled correctly, would be to loop over the statuses in the `WorkItemStatus` enum and ensure that one item per status exists, then the test could assert that all statuses end up in the correct bucket.

bilts reviewed (approved):
Minor comments.  None are strictly necessary for a merge, though, but they may offer some small improvements.

indiejames commented on test/workflow-orchestration.ts:
```diff
@@ -138,6 +138,70 @@ describe('Workflow chaining for a collection configured for swot reprojection an
+    hookRangesetRequest('1.0.0', collection, 'all', { query: reprojectAndZarrQuery });
+    hookRedirect('joe');
+
+    describe('when checking for a query-cmr work item', function () {
```
It's a little confusing that you are testing to see if query-cmr gets executed here instead of just executing it to get the side effect that you want (creation of the work items for the next step). In fact, query-cmr is not really special and should have the same behavior as any other service when the job is canceled (shouldn't get any work items), so you could just use it instead of a subsequent service. Not a hug deal, just a bit confusing to see a 'success' test case inside a 'job fails' `describe`.

indiejames commented on test/workflow-orchestration.ts:
```diff
@@ -138,6 +138,70 @@ describe('Workflow chaining for a collection configured for swot reprojection an
+          // job failure should trigger cancellation of any pending work items
+          const currentWorkItems = (await getWorkItemsByJobId(db, workItem.jobID)).workItems;
+          expect(currentWorkItems.length).to.equal(4);
+          expect(currentWorkItems.filter((i) => i.status === WorkItemStatus.SUCCESSFUL && i.serviceID === 'harmonyservices/query-cmr:latest').length).to.equal(1);
```
nit: prefer `item` to `i`

indiejames commented on test/workflow-orchestration.ts:
```diff
@@ -138,6 +138,70 @@ describe('Workflow chaining for a collection configured for swot reprojection an
+        });
+
+        describe('when checking to see if further reprojection work is queued', function () {
+          hookGetWorkForService('sds/swot-reproject:latest');
```
style nit: Feel free to disagree, but I dislike using hooks for a single test (probably have done it myself, though). I think it would be clearer if you just put a call to `getWorkForService` in the `it`. For me hooks make it harder to know what is going on and I only want to use them if they save a lot of redundant code.

indiejames replied:
Actually, looking at the code above (line 177) it would make sense to move this hook up so both of these test cases are using it since they are making the same call.

indiejames reviewed (changes requested):
Only one real  change request regarding use of `hookGetWorkForService` in test/workflow-orchestration.ts.

vinnyinverso commented:
@bilts @indiejames 

I think the most recent commit addresses all of the comments.

For the tests, I moved a bunch of logic into the describe blocks and cleaned things up.

For the actual implementation, I did my best to adhere to Patrick's suggestion by adding a helper function `completeJob` in the util/job module. Also, within the `updateWorkItem` route, I had to move the invocation of `updateWorkItemStatus()` after the block that checks whether its job is failed or canceled, to prevent a work item from updating its status (from cancelled to something else) after the job has failed or been canceled.
